crates/pglt_lsp/tests/server.rs

```diff
@@ -0,0 +1,341 @@
+                    .params(to_value(&params).context("failed to serialize params")?)
+                    .finish(),
+            )
+            .await
+            .map_err(Error::msg)
+            .context("call() returned an error")?
+            .map(|res| {
+                let (_, body) = res.into_parts();
+
+                let body =
+                    body.with_context(|| format!("response to {method:?} contained an error"))?;
+
+                from_value(body.clone()).with_context(|| {
+                    format!(
+                        "failed to deserialize type {} from response {body:?}",
+                        type_name::<R>()
+                    )
+                })
+            })
+            .transpose()
+    }
+
+    /// Basic implementation of the `initialize` request for tests
+    // The `root_path` field is deprecated, but we still need to specify it
+    #[allow(deprecated)]
+    async fn initialize(&mut self) -> Result<()> {
+        let _res: InitializeResult = self
+            .request(
+                "initialize",
+                "_init",
+                InitializeParams {
+                    process_id: None,
+                    root_path: None,
+                    root_uri: Some(url!("")),
+                    initialization_options: None,
+                    capabilities: ClientCapabilities::default(),
+                    trace: None,
+                    workspace_folders: None,
+                    client_info: None,
+                    locale: None,
+                },
+            )
+            .await?
+            .context("initialize returned None")?;
+
+        Ok(())
+    }
+
+    /// Basic implementation of the `initialized` notification for tests
+    async fn initialized(&mut self) -> Result<()> {
+        self.notify("initialized", InitializedParams {}).await
+    }
+
+    /// Basic implementation of the `shutdown` notification for tests
+    async fn shutdown(&mut self) -> Result<()> {
+        self.service
+            .ready()
+            .await
+            .map_err(Error::msg)
+            .context("ready() returned an error")?
+            .call(Request::build("shutdown").finish())
+            .await
+            .map_err(Error::msg)
+            .context("call() returned an error")
+            .and_then(|res| {
+                if let Some(res) = res {
+                    bail!("shutdown returned {:?}", res)
+                } else {
+                    Ok(())
+                }
+            })
+    }
+
+    async fn open_document(&mut self, text: impl Display) -> Result<()> {
+        self.notify(
+            "textDocument/didOpen",
+            DidOpenTextDocumentParams {
+                text_document: TextDocumentItem {
+                    uri: url!("document.js"),
+                    language_id: String::from("javascript"),
+                    version: 0,
+                    text: text.to_string(),
+                },
+            },
+        )
```

Comment:
Just a feeling: I think we need to be careful about what we abstract away and what we force the tests to do themselves.

For methods that are always called like `shutdown`, `initialize` etc. I think it makes sense to provide convenience methods.

But I'm not super about e.g. `open_document` – a bug might slip in because our testing logic contains too much "business logic". 

Also, this is javascript 🤣 

But yeah, WDYT? I think it's a fine line